### PR TITLE
plat/xilinx: Add support for new platform management APIs for ZynqMP

### PR DESCRIPTION
**Pull request by jollysxilinx**
This patch  adds support for new APIs to communication layer with firmware.
APIs include clock control, pinmux control, query and dev ioctl.

QUERY api provide interface to query soc specific data from ATF
IOCTLs provide interface to support different device configurations
Clock control APIs enables support for clock control using PMU(platform management unit)
Pinmux control APIs enables support for pin control using PMU.(platform management unit)

Signed-off-by: Jolly Shah jollys@xilinx.com

### CODE REVIEW

File: plat/xilinx/zynqmp/pm_service/pm_api_clock.c [hunk 1]
```diff
@@ -0,0 +1,3171 @@
...
+
+#define EXT_CLK_MIO_DATA(mio)				\
+	[EXT_CLK_INDEX(EXT_CLK_MIO##mio)] = {		\
+		.name = "mio_clk_"#mio,			\
+	}
+
+#define EXT_CLK_INDEX(n)	(n - CLK_MAX_OUTPUT_CLK)
+
+/* Clock control related definitions */
+#define BIT_MASK(x, y) (((1U << (y)) - 1) << (x))
```
Comment (sivadur) [inline plat/xilinx/zynqmp/pm_service/pm_api_clock.c]: can use BIT(y)

**Comment by davidcunado-arm** [comment]
@jollysxilinx 
Please confirm that additional IP belongs to xilinx?

@sivadur 
Please ACK this PR?

File: plat/xilinx/zynqmp/pm_service/pm_api_clock.c (hunk 1 above)
Comment (jollysxilinx) [reply plat/xilinx/zynqmp/pm_service/pm_api_clock.c]: Hi @sivadur ,
Above macro provides mask at offset x with y width so can not be replaced with BIT(y)

**Comment by jollysxilinx** [comment]
Hi @davidcunado-arm ,
yes it is xilinx specific.

**Comment by davidcunado-arm** [comment]
jenkins: test this please

**Comment by davidcunado-arm** [comment]
@jollysxilinx 
Looks like we're getting a build error in debug mode - the BL31 size is being exceeded. 

aarch64-linux-gnu-ld: build/zynqmp/debug/bl31/bl31.elf section `coherent_ram' will not fit in region `RAM'
aarch64-linux-gnu-ld: BL31 image has exceeded its limit.
aarch64-linux-gnu-ld: region `RAM' overflowed by 1 bytes
make: *** [build/zynqmp/debug/bl31/bl31.elf] Error 1

The build options we're using are:

DEBUG=1 PLAT=zynqmp RESET_TO_BL31=1

Can you check the map layout?

**Comment by jollysxilinx** [comment]
Will open a new pull request with a fix

**Comment by davidcunado-arm** [comment]
@jollysxilinx 
You don't need to open a new PR - just force-push to the branch on your fork and github will pick up the change and update the PR.

**Comment by jollysxilinx** [comment]
Sorry I closed it already and doesn't allow me to reopen now.

**Pull request closed without merging**